### PR TITLE
Implement basictracer for Javascript

### PR DESCRIPTION
**Pull request by CatTail**

### CODE REVIEW

**Comment by bhs** [comment]
Can you do something like https://github.com/opentracing/opentracing-javascript/blob/master/.gitignore before we get started? Thanks.

**Comment by bhs** [comment]
(fine to squash the original commit)

**Comment by CatTail** [comment]
@bensigelman Just update `.gitignore`, is this what you mean?

**Comment by bhs** [comment]
@CatTail yes; I was also suggesting that you squash your commits (via `git rebase -i` or similar) so the `dist/` files never end up in the repo, but that is not important.

File: src/tracer.js [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+        this._sampler = sampler || new Sampler()
+        this._recorder = recorder || new Recorder()
+        this._binaryPropagator = new BinaryPropagator(this)
+        this._textPropagator = new TextMapPropagator(this)
+    }
+    setInterface(inf) {
+        this._interface = inf
+    }
+    /**
+     * Starts and returns a new Span representing a logical unit of work.
```
Comment (bhs) [inline src/tracer.js]: nit: no need for the repeats of OpenTracing comments... it's probably better to just include a pointer in a `BasicTracer` class comment to the OpenTracing-js Tracer spec.

File: src/tracer.js [hunk 2]
```diff
@@ -0,0 +1,130 @@
+'use strict'
+
+import BasicSpan from './span'
+import {TextMapPropagator, BinaryPropagator} from './propagation'
+import Sampler from './sampler'
+import Recorder from './recorder'
+
+export default class BasicTracer {
+    constructor({sampler, recorder} = {}) {
```
Comment (bhs) [inline src/tracer.js:24]: please do include comments for the constructor and the expected types of the sampler and recorder

File: src/tracer.js [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+        } else if (format === this._interface.FORMAT_BINARY) {
+            span = this._binaryPropagator.join(operationName, carrier)
+        }
+        return span
+    }
+
+    isSample(span, parent) {
+        return this._sampler.isSample(span, parent)
+    }
+    record(span) {
```
Comment (bhs) [inline src/tracer.js]: should this be private?

File: src/tracer.js [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+        let span
+        if (format === this._interface.FORMAT_TEXT_MAP) {
+            span = this._textPropagator.join(operationName, carrier)
+        } else if (format === this._interface.FORMAT_BINARY) {
+            span = this._binaryPropagator.join(operationName, carrier)
+        }
+        return span
+    }
+
+    isSample(span, parent) {
```
Comment (bhs) [inline src/tracer.js]: why is the parent needed?

Comment (bhs) [inline src/tracer.js]: also: should this be private / `_isSampled`?

File: src/state.proto [hunk 5]
```diff
@@ -0,0 +1,8 @@
+syntax = "proto3";
+
+message TracerState {
```
Comment (bhs) [inline src/state.proto:3]: this really should be brought in through a submodule or something, but let's leave it this way for now.

File: src/index.js [hunk 6]
```diff
@@ -0,0 +1,10 @@
+'use strict'
+
+import tracer from 'opentracing'
+import Backend from './tracer'
+
+tracer.configure = (options) => {
+    tracer.initGlobalTracer(new Backend(options))
```
Comment (bhs) [inline src/index.js:14]: Sorry, where is Backend defined?

**Comment by CatTail** [comment]
@bensigelman 

> why is the parent needed?

In code

``` js
isSample(span, parent) {
}
```

Sometimes we need to decide whether current span should be used based on parent(carrier) information.

File: src/index.js (hunk 6 above)
Comment (CatTail) [inline src/index.js:14]: src/tracer.js is the `Backend`, and the name coming from https://github.com/opentracing/opentracing-javascript/blob/master/README.md example

``` js
var Tracer = require('opentracing');
var TracingBackend = require('tracing-implementation-of-your-choice');

Tracer.initGlobalTracer(TracingBackend.create());
```

File: src/sampler.js [hunk 7]
```diff
@@ -0,0 +1,8 @@
+export default class DefaultSampler {
+    isSample(span, parent) {
```
Comment (bhs) [inline src/sampler.js]: per the previous discussion (which is lost due to the squash): IMO it's not necessary to consider whether a non-root span (i.e., a span with a parent) should be sampled in BasicTracer... every tracing system I'm aware of just inherits the parent's sampling status in that case.

Also, `s/isSample/isSampled/`... and please add a brief comment explaining the semantics and expected parameter and return types.

File: src/recorder.js [hunk 8]
```diff
@@ -0,0 +1,17 @@
+export default class Recorder {
```
Comment (bhs) [inline src/recorder.js]: Please add a good comment for `Recorder` and `record`, as this is the most important public concept in `basictracer`.

File: src/recorder.js [hunk 9]
```diff
@@ -0,0 +1,17 @@
...
+    record(span) {
+        console.log(JSON.stringify({ // eslint-disable-line
```
Comment (bhs) [inline src/recorder.js:24]: something like a `DebuggingRecorder` would be great, but the default Recorder should do nothing at all.

File: src/span.js [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+            timestamp,
+        })
+    }
+
+    finish(finishTime = Date.now()) {
+        this.duration = finishTime - this.startTime
+        this._tracer._record(this)
+    }
+
+    static generateUUID() {
```
Comment (bhs) [inline src/span.js:78]: I'm trusting you that this is the right way to generate a random int64 in javascript :)

File: src/propagation.js [hunk 11]
```diff
@@ -0,0 +1,108 @@
...
+const FIELD_NAME_SAMPLED = PREFIX_TRACER_STATE + 'sampled'
+const FIELD_COUNT = 3
+
+class TextMapPropagator {
+    constructor(tracer) {
+        this._tracer = tracer
+    }
+
+    inject(span, carrier) {
+        carrier[FIELD_NAME_TRACE_ID] = span.traceId
```
Comment (bhs) [inline src/propagation.js]: it's important for this to do an _unsigned_ int64->string conversion (same with below, and with the other direction in join).

File: src/__tests__/tracker-test.js [hunk 12]
```diff
@@ -0,0 +1,76 @@
...
+        let span = tracer.join(ANOTHER_OPERATION_NAME, FORMAT_TEXT_MAP,
+                               carrier)
+        should(span.traceId.equals(parentSpan.traceId)).be.ok()
+        should(span.spanId.equals(parentSpan.spanId)).be.not.ok()
+        should(span.parentId.equals(parentSpan.spanId)).be.ok()
+        should(span.sampled).eql(parentSpan.sampled)
+        should(span.baggage).eql(parentSpan.baggage)
+    })
+
+    it('should join binary span', () => {
```
Comment (bhs) [inline src/__tests__/tracker-test.js]: thanks for these tests, btw.

File: Readme.md [hunk 13]
```diff
@@ -0,0 +1,56 @@
+# basictracer-javascript [![NPM version][npm-image]][npm-url] [![build status][travis-image]][travis-url] [![Test coverage][coveralls-image]][coveralls-url]
```
Comment (bhs) [inline Readme.md:1]: I have a bunch of thoughts about the README, but as a thank-you to you for doing all of this work, I can just make README changes in my own followup PR :)

**Comment by CatTail** [comment]
@bensigelman Just push a new commit, any problem?

File: src/sampler.js [hunk 14]
```diff
@@ -0,0 +1,9 @@
+export class DefaultSampler {
+    /**
+     * @param  {Span} span Span to be determined.
```
Comment (bhs) [inline src/sampler.js:3]: BasicSpan or opentracing Span?

File: src/tracer.js [hunk 15]
```diff
@@ -0,0 +1,73 @@
...
+        this._textPropagator = new TextMapPropagator(this)
+    }
+
+    setInterface(inf) {
+        this._interface = inf
+    }
+
+    startSpan(fields) {
+        // Interface or Implementation argument
+        // https://github.com/opentracing/opentracing-javascript/pull/29
```
Comment (bhs) [inline src/tracer.js:38]: it didn't look like https://github.com/opentracing/opentracing-javascript/pull/29 was going to go through... isn't it safe to assume that `fields.parent` is an opentracing `Span` and thus has an `imp()` method?

**Comment by bhs** [comment]
Basically LGTM! Thanks. I have some nits about comments and documentation but it makes more sense for me to just suggest those changes separately, I think.

We should resolve the question about https://github.com/opentracing/opentracing-javascript/pull/29 and merge. Thanks again.

File: src/tracer.js (hunk 15 above)
Comment (CatTail) [inline src/tracer.js:38]: Maybe, but opentracing library don't expose it's `Tracer` class, so there is no way to do that

**Comment by CatTail** [comment]
@bensigelman Just add a little comment for `Recorder`, but I'm not a native english speaker, my comment maybe not appropriate (accurate)

File: src/tracer.js (hunk 15 above)
Comment (bcronin) [inline src/tracer.js:38]: I'm a little confused as well. I would have thought that `fields.parent` is _always_ a valid OpenTracing Span "Interface" object in the call to `startSpan`. Therefore, line 44 could simply be `parent: fields.parent.imp()` and line 39-41 could be removed. I'm not sure I understand how `Tracer` is involved.

My apologies if I'm missing something obvious here!

**Comment by bhs** [comment]
@CatTail thanks for the updates. I'm curious to hear your response to @bcronin's most recent comment (about `imp()`).

**Comment by bcronin** [comment]
This looks like a lot of good code. Thanks @CatTail .

One issue / question that can possibly be addressed _after_ this pull request:

It looks like this code is currently Node-specific (for example, it uses `Buffer`).  How difficult would it be to make this code work both on the browser and in Node?  One option would be to rename this repository `basictracer-node` and keep it Node-specific. Another option would be to make it "universal" so that it works on both the browser and Node.  I personally don't have a strong opinion on which approach is better.  In either case, I think this is a good starting point and browser-compatibility could be addressed in future changes.

File: src/tracer.js (hunk 15 above)
Comment (CatTail) [inline src/tracer.js:38]: You are right, the only case is in [unittest](https://github.com/CatTail/basictracer-javascript-1/blob/4a50f7b1d163126df6b5bd275d662137f0d1c0a1/src/__tests__/tracker-test.js#L68). I'm going to change the unittest.

**Comment by CatTail** [comment]
@bcronin There are work need to be done to use in browser side. How about figure it out in future version?

**Comment by CatTail** [comment]
@bcronin @bensigelman I think the bridge pattern limit the implementation's ability to extend. For example, how can I add a format support HTTP format?

Text format is not suitable because add `x-` prefix for HTTP header is more appropriate. Checkout [my refactor version of basictracer](https://github.com/CatTail/basictracer-javascript/blob/0a73ec2a190268a4b54aa26c9e05bb9b93476173/src/propagation.js#L49), it directly expose tracer rather use bridge pattern.

**Comment by bhs** [comment]
@CatTail FYI, the `X-` prefix has been deprecated since 2012: https://tools.ietf.org/html/rfc6648

Regardless, though, the idea with the "carriers" is that they can be adapted to HTTP headers or to other key:value maps (for example, grpc.io "metadata": https://github.com/grpc/grpc/blob/release-0_13/src/node/src/metadata.js) without requiring any additional work from Tracer implementations.

File: src/tracer.js [hunk 16]
```diff
@@ -0,0 +1,74 @@
...
+import {DefaultSampler} from './sampler'
+import {DefaultRecorder} from './recorder'
+
+// Implement https://github.com/opentracing/opentracing-javascript/blob/master/src/tracer.js
+export default class BasicTracer {
+    /**
+     * Constructs basic tracer instance with provided Sampler and Recorder
+     * @param  {Object} options
+     *         Optional associative array of fields.
+     *         - `sampler` {Sampler} Optional object with `isSample` method, the
```
Comment (bhs) [inline src/tracer.js:14]: nit: "isSampled" in the comment

File: src/index.js [hunk 17]
```diff
@@ -0,0 +1,20 @@
...
+
+import tracer from 'opentracing'
+import Backend from './tracer'
+import {DefaultSampler} from './sampler'
+import {DefaultRecorder, DebuggingRecorder} from './recorder'
+
+
+// Little hack because interface don't expose a way to extend supported
+// carrier format
+tracer.FORMAT_HTTP_HEADER = 'http_header'
```
Comment (bhs) [inline src/index.js:11]: I don't understand why this is here... can you explain?

**Comment by bcronin** [comment]
@CatTail regarding the refactor that "directly expose tracer rather use bridge pattern.":

My concern here (and I'm glad to be corrected if I am missing something) is this removes what seems to me to be a necessary level of indirection. It would be great if third-party packages (for example, express.js) could be instrumented with OpenTracing _but not tied to an implementation_. That means that express.js package code needs to refer to an interface only.  If I then write my web application, let's call it MyApp, that _uses_ express.js, I need a way to connect express.js' instrumented code to the _implementation_ that MyApp is using.

This could be done with true global variables (using the `global` built-in variable) but there are problems with that (namely, many JS developers would consider it a bad practice and unit testing tools like `mocha` will complain by default about this). It could also be done by adding the tracer implementation to the initialization of express.js, but that would mean every single OpenTracing-compatible package needs to modify their initialization API and each one needs to be passed that implementation object  (and continue to pass it to its dependencies in turn); that is certainly _possible_, but is inconvenient and error-prone, and also quite awkward in utility libraries that don't have a specific initialize call in their API.

If there's a good way to instrument a third-party package with just the interface and then, in an uncomplicated way, connect (at runtime) a package to the host application's tracing implementation, I would enjoy hearing it and talking about alternatives to the bridge pattern.

File: src/index.js (hunk 17 above)
Comment (CatTail) [inline src/index.js:11]: Because `opentracing-javascript` don't have a way to add additional format

**Comment by CatTail** [comment]
@bcronin I don't have any good idea to restrict implementation API. Compared to [basictracer-python](https://github.com/opentracing/basictracer-python), they use inheritance to defined API and extend the ability to allow implementation to do more, such as support more carrier format.

**Comment by bcronin** [comment]
@CatTail , I think I partly understand what you are saying. My understanding is that `FORMAT_HTTP_HEADER` is _not_ part of the OpenTracing standard; therefore, I'd argue it _should not_ be part of the `opentracing` package / namespace. From that perspective, I see it as potentially a good thing that a custom format can't be added to the set of constants exposed by the `opentracing` package. 

That said, the format argument to `join` and `inject` is defined as a `string`. Nothing prevents the implementation package from defining a new constant and using that. I don't fully understand how the current code prevents more carrier formats from being defined. They won't be under the same package / namespace, but again, I think that is correct since they are (by definition) not part of the standard.

Apologies if I am misunderstanding what you are saying!

**Comment by bhs** [comment]
I think @bcronin is right about the supported formats... I think we should just merge this and make changes as followups. It's already in a useful place and is much better than what's at `master` (which is still empty).

I'll create some issues to track things that still need to be done.

**Comment by bhs** [comment]
Issues opened: #2 #3 #4

**Pull request merged**